backend/controllers/base.go

```diff
@@ -0,0 +1,92 @@
+package controllers
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"net/http"
+	"peoplemath/storage"
+)
+
+// ObjectUpdateResponse is returned to the browser after an insert or update (e.g. for concurrency control)
+type ObjectUpdateResponse struct {
+	LastUpdateUUID string `json:"lastUpdateUUID"`
+}
+
+type BaseController struct {
+	Store storage.StorageService
+}
+
+type TeamController struct {
+	*BaseController
+}
+
+type PeriodController struct {
+	*BaseController
+}
+
+type ImproveController struct {
+	*BaseController
+}
+
+var (
+	improveController *ImproveController
```

Comment:
My idea was to implement MVC (https://ru.wikipedia.org/wiki/Model-View-Controller) structure, but in our case only Model/Controller layers are acceptable. I agree with you if you do not want it - it is just structure changes. From my point of view, Controllers layer is a special case of the implementation of the principle of single responsibility - each layer does its own thing: Controllers layer works with the client request at the top level, it can delegate data processing to models; it can have separate from other controllers functionality (for example you have methods to get request vars from URL - they are different for each URL, in my implementation I put them separate to each controller) and also it have have common methods shared in BaseController.

Reply:
I see what you are driving at. It's definitely nice to keep the model and controllers separate; I did have separate data structures for these in my original implementation, which you then moved into separate packages, which was a very nice improvement.

As you noted, we don't have any sort of view logic in the backend: that lives in the Angular frontend.

In the current design, the `Server` struct is a collection of controller methods. Essentially your proposal doesn't actually introduce any new architectural layers, it just splits the `Server` into separate structs for the different HTTP endpoints. If we did that then we could get rid of the `Server` altogether, as it wouldn't have any logic left in it.

At the moment, my sense is that such a split wouldn't justify the added complexity of sharing the `StorageService` and other `Server` members, plus the shared logic, between the controllers. I think we can get a lot of the benefit simply by splitting the `Server` methods into different files, one for each HTTP endpoint. That way we get shorter files with a logical separation of functionality, but retain a simple design.